### PR TITLE
core: check for pointer equality when comparing Eq slices

### PR DESCRIPTION
**Pull request by aschampion**
Because `Eq` types must be reflexively equal, an equal-length slice to the same memory location must be equal.

This is related to #33892 (and #32699) answering this comment from that PR:

> Great! One more easy question: why does this optimization not apply in the non-BytewiseEquality implementation directly above?

Because slices of non-reflexively equal types (like `f64`) are not equal even if it's the same slice. But if the types are `Eq`, we can use this same-address optimization, which this PR implements. Obviously this changes behavior if types violate the reflexivity condition of `Eq`, because their impls of `PartialEq` will no longer be called per-item, but 🤷‍♂ .

It's not clear how often this optimization comes up in the real world outside of the same-`&str` case covered by #33892, so **I'm requesting a perf run** (on MacOS today, so can't run `rustc_perf` myself). I'm going ahead and making the PR on the basis of being surprised things didn't already work this way.

This is my first time hacking rust itself, so as a perf sanity check I ran `./x.py bench --stage 0 src/lib{std,alloc}`, but the differences were noisy.

To make the existing specialization for `BytewiseEquality` explicit, it's now a supertrait of `Eq + Copy`. `Eq` should be sufficient, but `Copy` was included for clarity.

### CODE REVIEW

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -5304,6 +5304,29 @@ impl<A, B> SlicePartialEq<B> for [A]
...
+    default fn equal(&self, other: &[A]) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+
+        if self.as_ptr() == other.as_ptr() {
+            return true;
+        }
+
+        for i in 0..self.len() {
```
Comment (Centril) [inline src/libcore/slice/mod.rs]: This could be written as:
```rust
        self.iter().zip(other.iter()).all(|(x, y)| x == y)
```

Comment (aschampion) [reply src/libcore/slice/mod.rs]: Good point. I mindlessly copied the other impl, but as written it's not clear the bounds check for `other[i]` would always be elided. I've updated both impls to user iterators now.

**Comment by aschampion** [comment]
`i128` and `u128` should also be `BytewiseEquality` so they can take advantage of the `memcmp` impl. Should I push that here rather than opening another PR for a 1 line diff?

**Comment by Centril** [comment]
@rust-timer build 261715133edddbeeccb64c5a890ce76658fe589f

**Comment by sfackler** [comment]
> `i128` and `u128` should also be `BytewiseEquality` so they can take advantage of the `memcmp` impl. Should I push that here rather than opening another PR for a 1 line diff?

Either way - a separate PR may be best to avoid accidentally dropping the change if the perf run reports that the pointer equality check isn't worth it though.

**Comment by Dylan-DPC** [comment]
ping from triage @sfackler @aschampion any updates on this?

**Comment by sfackler** [comment]
The benchmark results seem like a pretty mixed bag - not sure if this change really had much of an impact.

**Comment by aschampion** [comment]
For context, I created this because I had code (change queue consistency check) where the most common path was same slice comparison, and having exploited the primitive slice equal pointer optimization before was surprised when the performance was linear and not constant. Of course, if you want to rely on this optimization it's easy to manually check for this case, which is what I've done in the meantime. However, it doesn't look like this comes up frequently in the benchmarking suite. I've looked through a few popular crates trying to find cases where this would apply, but haven't found any yet.

So it's down to whether having behavior consistent with primitive slices is seen as more consistent/less surprising, or if having another `impl` is noise for a rare niche optimization, in which case this can be closed. I can sympathize with both views.

**Comment by sfackler** [comment]
Yeah, that all makes sense - seems like it's worth discussing at the next libs triage meeting.

**Comment by sfackler** [comment]
We talked about this at the libs meeting today, and tentatively felt like we should land this. It doesn't appear to regress normal code in a significant way, and we can always back it out later!

@bors r+

**Pull request merged**